### PR TITLE
Added generic equation functionality

### PR DESCRIPTION
**Pull request by nnooney**
This is a fix of issue #10. See the example demos/cartesian.html. I've also implemented a module that will run the Marching Squares algorithm or the Marching Cubes algorithm. Currently it only supports the Marching Squares algorithm in 2d; will get to implementing the Marching Cubes algorithm in the future.

### CODE REVIEW

**Comment by nnooney** [comment]
Also, the demo added a new gui component, resolution. This value represents the number of "squares" to divide the graph into. The default value is 20, which results in a resolution per integer.

File: demos/cartesian.html [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	    var viewing = gui.addFolder('Viewing Controls');
+
+	    // Add shape options
+	    // Should be able to: (1) Draw from cartesian equation (2) Draw from points (3) Draw from parametric equation
+	    shapeProperties.add(params, 'equation').name('Equation').onChange(function(val){
+	    	// Check that the equation is correct first
+	    	try {
+	    		// If that succeeds, go ahead
+	    		equation.parse(val);
+					renderGraph();
```
Comment (sambrodriguez) [inline demos/cartesian.html:97]: minor: extra whitespace <3

Comment (nnooney) [reply demos/cartesian.html:97]: While this is true, the whole demos section needs reorganization in the future.

**Comment by OmarShehata** [comment]
Very cool! Is Polygonize a library or you just put that together? 

I noticed the intersection doesn't work anymore. I made a commit that makes it work for the Y axis (looks really cool on a circle!) though couldn't get it to work on the X axis intersection. 

Also, instead of naming your branch as your username, we're going with the convention of naming it based on the feature it implements, and then closing and deleting the branch once its merged. 

Feel free to review the commit I made and merge if you don't have any other edits to make now!

**Comment by OmarShehata** [comment]
Another question to consider: Can the marching squares draw a filled in region? How would we go about doing that?

**Comment by nnooney** [comment]
I actually wrote it mimicing the style of the Parser.js file and looking at two examples (one for marching squares and one for marching cubes). I wrote the code, which is very similar to those algorithms used. Both are described here:

http://paulbourke.net/geometry/polygonise/

As for filling in regions with marching squared - I believe with some effort it could. What I return from the algorithm is a series of line segments along the line where I believe the equation is equal to zero. All that mathbox does is render those line segments.

I could hook into the algorithm such that when it encounters a cell that should be filled in entirely, then it could call another function (fillCell) that would generate the triangles to fill in the square. This is sufficient in 2D, but it would not scale well to 3D.

Another option would be to use the line segments returned from the algorithm to look at the regions of the graph that it divides (in many examples this divides the visible area into two sections). Then with the division of the graph, you could test a point in each section to determine whether or not it should be shaded and draw another layer on that graph for the fill. A fill-cell algorithm would work well here, or perhaps a scanline algorithm since we aren't guaranteeing convex shapes.

To keep the isolation of functionality separate, I think the second option would be better to implement. Maybe this requires exposing some of the stuff in Polygonize.

**Comment by OmarShehata** [comment]
Well, if your algorithm returns a set of points that make up the outline of the polygon, and whether or not it's convex, couldn't we just run a triangulation algorithm to convert it into a set of triangles, and draw those with mathbox? That way functionality remains separate and it works for any shape. 

In terms of 3D, if marching cubes gives you the same kind of data, can we then just draw it with something like Threejs's [Convex Geometry object](https://github.com/StoDevX/humke-4d-geometry/blob/master/demos/threejs_convex_geometry.html) ? (I guess we'd need a way to ensure it's convex then)

**Pull request merged**